### DISCUSSION TITLE
Persist router stack when activity is killed by android os

### DISCUSSION THREAD

**Discussion by Lingareddyvinitha** [csells/go_router#328] [Q&A]
I am currently facing the issue, that (if RAM is low etc.) Android may kill the Flutter Activity. I can persist the state of my widgets etc. with no problem, but the current route history always is lost and the app starts again at the initial route.

Is there some way to persist the current stack of the router?

**Comment by csells** [comment]
go_router supports state restoration. Here's an example that shows off how that works: https://github.com/csells/go_router/blob/main/go_router/example/lib/state_restoration.dart

**Reply by Lingareddyvinitha** [reply]
> Thanks for your answer. Here the problem is not with state of the current route.
When the activity is killed by android os current route is persisting, then **when we press the mobile back button instead of going to the previous screen, app is closing**

**Reply by csells** [reply]
@lulupointu thoughts here?

**Reply by lulupointu** [reply]
I just tried with the default example, modified so that the back button pops (see code bellow). I can't reproduce the issue.

Here is what I have done:
  1. Enable "Don't keep activities" in the developer settings of Android
  2. Launch the flutter app (code bellow)
  3. Click on "Go to page 2"
  4. The Page2 is shown
  5. Hit the home button
  6. Click the app icon
  7. Wait for the app to reload (hint that "Don't keep activities" worked)
  8. See the app being on page 2 (hint that "State restoration" worked)
  9. Hit the android back button
  10. See the Page 1

If I understand the issue correctly, at step 10 the application would have close if there was any issue. However on my machine it did pop on the Page 1 which is what is expected.
One thing to keep in mind is that I did *change the original example* to nest the routes:

```dart
// DO
GoRoute(
  path: '/',
  builder: (context, state) => const Page1Screen(),
  routes: [
    GoRoute(
      path: 'page2',
      builder: (context, state) => const Page2Screen(),
    ),
  ],
)

// DON'T
GoRoute(
  path: '/',
  builder: (context, state) => const Page1Screen(),
),
GoRoute(
  path: '/page2',
  builder: (context, state) => const Page2Screen(),
),
```

However this has nothing to do with state restoration or closing the app, it's just the way `GoRouter` always work.

@Lingareddyvinitha feel free to tell me if I missed something

<details>
<summary>
Click to reveal code
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() => runApp(
      const RootRestorationScope(restorationId: 'root', child: App()),
    );

class App extends StatefulWidget {
  const App({Key? key}) : super(key: key);

  static const title = 'GoRouter Example: State Restoration';

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> with RestorationMixin {
  @override
  String get restorationId => 'wrapper';

  @override
  void restoreState(RestorationBucket? oldBucket, bool initialRestore) {
    // todo: implement restoreState for you app
  }

  @override
  Widget build(BuildContext context) => MaterialApp.router(
        routeInformationParser: _router.routeInformationParser,
        routerDelegate: _router.routerDelegate,
        title: App.title,
        restorationScopeId: 'app',
      );

  final _router = GoRouter(
    routes: [
      // restorationId set for the route automatically
      GoRoute(
        path: '/',
        builder: (context, state) => const Page1Screen(),
        routes: [
          // restorationId set for the route automatically
          GoRoute(
            path: 'page2',
            builder: (context, state) => const Page2Screen(),
          ),
        ],
      ),
    ],
    restorationScopeId: 'router',
  );
}

class Page1Screen extends StatelessWidget {
  const Page1Screen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: const Text(App.title)),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              ElevatedButton(
                onPressed: () => context.go('/page2'),
                child: const Text('Go to page 2'),
              ),
            ],
          ),
        ),
      );
}

class Page2Screen extends StatelessWidget {
  const Page2Screen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: const Text(App.title)),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              ElevatedButton(
                onPressed: () => context.go('/'),
                child: const Text('Go to home page'),
              ),
            ],
          ),
        ),
      );
}
```

</details>

**Reply by Lingareddyvinitha** [reply]
Thanks for your answer. It's working
